src/autoBatchNode.js

```diff
@@ -0,0 +1,86 @@
+
+const Endpoint = require('./endpoint')
+const _ = require('lodash')
+
+// This helper function returns all method names including those inherited from extending classes
+function getClassMethodNames(klass) {
+  const isFunction = (x, name) => typeof x[name] === 'function';
+  const deepFunctions = x =>
+      x !== Object.prototype &&
+      Object.getOwnPropertyNames(x)
+          .filter(name =>  isFunction(x, name))
+          .concat(deepFunctions(Object.getPrototypeOf(x)) || []);
+  const distinctDeepFunctions = klass => Array.from(new Set(deepFunctions(klass)));
+  return distinctDeepFunctions(klass).filter(name => name !== 'constructor' && !name.startsWith('__'))
+}
+
+module.exports = class AutoBatchNode {
+  constructor(parent, client) {
+    this.parent = parent
+    this.client = client
+    this.children = {}
+    
+    getClassMethodNames(parent).forEach((key) => {
+      this[key] = (...args) => {
+        // Enforce settings from the Client
+        this._syncSettingsFromClient()
+
+        // Return endpoint node from the child pool, if available
+        const child = this._getChildNode(key, args)
+        if (child) {
+          return child
+        }
+
+        // Call the parent's method
+        const result = parent[key](...args)
+
+        // If the method returns parent, return this
+        if (result === this.parent) {
+          return this
+        } 
+        // If method returns a new Endpoint, add it to the child pool and return it
+        else if (result instanceof Endpoint) {
+          return this._newChildNode(key, args, result)
+        }
+        // Otherwise, return the result of the method
+        else {
+          return result
+        }
+      }
+    })
+  }
+
+  _syncSettingsFromClient() {
+    // If this is the client, don't need to change settings
+    if (this.parent === this.client) {
+      return 
+    }
+    // Copy the settings from client to this endpoint
+    this.parent
+      .schema(this.client.schemaVersion)
+      .language(this.client.lang)
+      .authenticate(this.client.apiKey)
+      .debugging(this.client.debug)
+    // Autobatching endpoints should always use cache
+    this.parent._skipCache = false
+
+    return this
+  }
+
+  _newChildNode(endpointName, args, endpoint) {
+    endpoint.enableAutoBatch()
+
+    const childNode = new AutoBatchNode(endpoint, this.client)
+    const nameList = [endpointName, ...args]
+    const childName = nameList.join('_')
+    this.children[childName] = childNode
+    return childNode
+  }
+
+  _getChildNode(endpointName, args) {
+    const nameList = [endpointName, ...args]
+    const childName = nameList.join('_')
+    return this.children[childName]   
+  }
+
+}
```

Comment:
I have to say the entire wrapping logic here feels a little over-engineered. What benefits does it have over just passing a `autoBatching = true` flag, just like we pass flags for e.g. `debug`?

- Set the debug flag on the client: https://github.com/queicherius/gw2api-client/blob/master/src/client.js#L45-L49
- Inherit the debug flag from the parent (this can be a client or a other endpoint): https://github.com/queicherius/gw2api-client/blob/master/src/endpoint.js#L16
- Set the debug flag on the endpoint: https://github.com/queicherius/gw2api-client/blob/master/src/endpoint.js#L57-L61
- Use the debug flag: https://github.com/queicherius/gw2api-client/blob/master/src/endpoint.js#L65

It feels like that would ultimately achieve the same functionality and would e.g. also make it possible to pass the `autoBatchDelay` to the client if we did something like this:

```js
// Client

autoBatch (flag = true, autoBatchDelay = 50) {
  this.autoBatching = flag
  this.autoBatchDelay = autoBatchDelay
  return this
}
```

```js
// Endpoint

constructor (parent) {
  // ...
  this.autoBatching = parent.autoBatching
  this.autoBatchDelay = parent.autoBatchDelay

  this._autoBatch = {
    idsForNextBatch: new Set(),
    nextBatchPromise: null,
  }
}

autoBatch (flag = true, autoBatchDelay = 50) {
  // We could also print the debug warning message here, if this.isBulk = false
  // that the flag will just be ignored. But in some cases it might make sense anyway,
  // like this one where it gets passed to a child:
  // client().account().autoBatch(true).characters().many([...])

  this.autoBatching = flag
  this.autoBatchDelay = autoBatchDelay
  return this
}

_many (ids, partialRequest = false, skipAutoBatch = false) {
  if (this.isBulk && this.autoBatching && !skipAutoBatch) {
    return this._autoBatchMany(ids)
  }
}
```

```js
// Usage
const api = client().autoBatch(true, 50)
api.items().many([12]) // -> autobatches
api.items().autoBatch(false).many([12]) // -> doesn't
client().items().autoBatch(true).get(12) // -> does
```

Reply:
`I` agree it would be cleaner if autobatch were handled via setting flag, however I think it could be confusing. For example:
`let api = client().autoBatch()
api.items().get(101)
api.items().get(102)`
This normal behavior of the library is to return a new endpoint from items(), which would have seperate autoBatch sets of ids with my current endpoint proposed edit.  
In my mind, if the client is set to autoBatch, then subsequent calls for the same endpoint should also autoBatch. The main thing the wrapper is doing is caching child endpoints so that subsequent calls are returning the same endpoint objects. For the most part, other method calls are passed through without interference.

The reason I added logic to re-inherit settings from the parent, was probably over-cautious, and unnecessary.  I was worried that people might try to change settings on the endpoint like this:
`await api.items().live().get(101)
await api.items().get(101)`
Which is fine with the current library because it makes a new endpoint on each line so the second line will use the cache. But if the api is returning saved endpoints from a pool for autobatching, then the setting stays.  This can be avoided with proper usage of the library, and removing it cuts some fluff from my edit.

Reply:
> This normal behavior of the library is to return a new endpoint from items(), which would have seperate autoBatch sets of ids with my current endpoint proposed edit. In my mind, if the client is set to autoBatch, then subsequent calls for the same endpoint should also autoBatch.

Ah! I didn't think about batching being shared between endpoints of the same type. That's very valid, and I agree that it should work how you described.

> The main thing the wrapper is doing is caching child endpoints so that subsequent calls are returning the same endpoint objects. For the most part, other method calls are passed through without interference.

Instead of wrapping the client and returning the same endpoint objects, what do you think about a singleton shared data object that holds the ID list and timeout based on the endpoint URL? That should solve the issues around the shared settings and still have the simple interface I suggested.

Something like this?

```js
const autoBatchSharedData = {}

module.exports = class AbstractEndpoint {
  constructor(parent) {
    // ...

    this.setupAutoBatchSharedData()
  }

  // Setup the shared data for auto batching
  setupAutoBatchSharedData() {
    const autoBatchId = this.baseUrl + this.url + ':' + this.schemaVersion
    if (!autoBatchSharedData[autoBatchId]) {
      autoBatchSharedData[autoBatchId] = {
        idsForNextBatch: new Set(),
        nextBatchPromise: null,
      }
    }

    this._autoBatch = autoBatchSharedData[autoBatchId]
  }

  // Set the schema version
  schema(schema) {
    this.schemaVersion = schema
    this.setupAutoBatchSharedData()
    this.debugMessage(`set the schema to ${schema}`)
    return this
  }
}
```

> The reason I added logic to re-inherit settings from the parent, was probably over-cautious, and unnecessary. I was worried that people might try to change settings on the endpoint like this:

```js
await api.items().live().get(101)
await api.items().get(101)
```

Yeah, I would also expect the first one to go through and the second one to hit the cache.

Reply:
I've been staring at this and wondering why I didn't think of that design before. 

Should language/apikey be added to the batchId with schema or remove schema altogether?
If we care, it should probably match _cacheHash. However this will only cause an issue if the user changes the setting inside of the batch window, which is 50ms by default... Probably just a result of irresponsible use, and could be a consideration in documentation about autoBatching

Reply:
> Should language/apikey be added to the batchId with schema or remove schema altogether?
If we care, it should probably match _cacheHash. However this will only cause an issue if the user changes the setting inside of the batch window, which is 50ms by default...

I think using `_cacheHash` is a good idea! Then it'd behave correctly in all instances (without race conditions) and the only overhead would be to add the extra function call to the setters for language & api key. Then if a user changes the settings in the batch window it doesnt get batched but it gets the correct result.